bin/docker-to-squash.py

```diff
@@ -0,0 +1,1430 @@
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+docker_to_squash.py is a tool to facilitate the process of converting
+Docker images into squashFS layers, manifests, and configs.
+
+Tool dependencies: skopeo, squashfs-tools, tar, setfattr
+"""
+
+import argparse
+from collections import Iterable
+import glob
+import hashlib
+import json
+import logging
+import os
+import re
+import shutil
+import subprocess
+from threading import Timer
+
+LOG_LEVEL = None
+HADOOP_BIN_DIR = None
+
+def shell_command(command, print_stdout, print_stderr, raise_on_error,
+                  timeout_sec=600):
+  global LOG_LEVEL
+  stdout_val = subprocess.PIPE
+  stderr_val = subprocess.PIPE
+
+  logging.debug("command: %s", command)
+
+  if print_stdout:
+    stdout_val = None
+
+  if print_stderr or LOG_LEVEL == "DEBUG":
+    stderr_val = None
+
+  process = None
+  try:
+    process = subprocess.Popen(command, stdout=stdout_val,
+                               stderr=stderr_val)
+    timer = Timer(timeout_sec, process_timeout, [process])
+
+    timer.start()
+    out, err = process.communicate()
+
+    if raise_on_error and process.returncode is not 0:
+      exception_string = ("Commmand: " + str(command)
+                          + " failed with returncode: "
+                          + str(process.returncode))
+      if out != None:
+        exception_string = exception_string + "\nstdout: " + str(out)
+      if err != None:
+        exception_string = exception_string + "\nstderr: " + str(err)
+      raise Exception(exception_string)
+
+  except:
+    if process and process.poll() is None:
+      process.kill()
+    raise Exception("Popen failure")
+  finally:
+    if timer:
+      timer.cancel()
+
+  return out, err, process.returncode
+
+def process_timeout(process):
+  process.kill()
+  logging.error("Process killed due to timeout")
+
+def does_hdfs_entry_exist(entry, raise_on_error=True):
+  out, err, returncode = hdfs_ls(entry, raise_on_error=raise_on_error)
+  if returncode is not 0:
+    return False
+  return True
+
+def setup_hdfs_dirs(dirs):
+  if does_hdfs_entry_exist(dirs, raise_on_error=False):
+    return
+
+  hdfs_mkdir(dirs, create_parents=True)
+  chmod_dirs = []
+  for dir_entry in dirs:
+    directories = dir_entry.split("/")[1:]
+    dir_path = ""
+    for directory in directories:
+      dir_path = dir_path + "/" +  directory
+      logging.info("dir_path: %s", str(dir_path))
+      chmod_dirs.append(dir_path)
+  hdfs_chmod("755", chmod_dirs)
+
+def append_or_extend_to_list(src, src_list):
```

Comment:
Parameter order here is counter-intuitive. It makes more sense if reversed.

Reply:
It means appending `src` to `src_list`. I think this order makes sense